### PR TITLE
Editorial for problem Hot & Cold

### PR DESCRIPTION
**Pull request by TheGamingMousse**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by TheGamingMousse** [comment]
https://github.com/cpinitiative/usaco-guide/issues/4165

File: solutions/platinum/bts-hotcold.mdx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		for (int v : adj[u]) {
+			if (v != p) { tour(v, u); }
+		}
+		tout[u] = timer - 1;
+	};
+	tour(1, 0);
+	tin[0] = -1, tout[0] = n + 1;  // so LCA works
+
+	// just some helper methods to simplify the code
+	auto isAnc = [&](int u, int v) -> bool {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:90]: use camlcase

File: solutions/platinum/bts-hotcold.mdx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		return tin[u] <= tin[v] && tout[v] <= tout[u];
+	};
+	auto lca = [&](int u, int v) -> int {
+		if (isAnc(u, v)) return u;
+		if (isAnc(v, u)) return v;
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:95]: braces

File: solutions/platinum/bts-hotcold.mdx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		return lift[0][u];
+	};
+	auto dist = [&](int a, int b, int anc = -1) -> int {
+		if (anc == -1) { anc = lca(a, b); }
+		return depth[a] + depth[b] - 2 * depth[anc];
+	};
+
+	// calculating difference array updates
+	vector<array<ll, 2>> diff(n + 1, {0, 0});
+	auto upd = [&](int u, int v, int initial, int change) -> void {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:108]: at this point i'd rather you put things in a class to make things more encapsulated

File: solutions/platinum/bts-hotcold.mdx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+				if (depth[split_1] < depth[split_2]) {
+					swap(split_1, split_2), swap(a, b);
+				}
+				upd(a, split_1, dist(a, t, split_1), -1);
+				upd(split_1, anc, dist(split_1, t, split_1), 1);
+				upd(b, lift[0][anc], dist(b, t, anc), -1);
+			}
+		}
+	};
+	while (s--) {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:163]: for loop

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
you're the only one on this site that gets me

File: solutions/platinum/bts-hotcold.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/platinum/bts-hotcold.mdx:108]: would this be like a Tree class where you put all the graph stuff in there?

Comment (SansPapyrus683) [reply solutions/platinum/bts-hotcold.mdx:108]: yeah, see my current pr for how it might go

File: solutions/platinum/bts-hotcold.mdx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+**Case 2: $p$ is equal to either $a$ or $b$.**
+
+Let the first ancestor of $t$ that is on the path from $a$ to $b$
+be $l$. WLOG, let $b$ be equal to $p$. Then, we update the path from
+$a$ to $l$ and the path from $l$ to $b$.
+
+**Case 3: The LCA of $t$ with $a$ and $b$ is the same as $p$.**
+
+Handle this case similarly to case 1.
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:29]: elaborate?

File: solutions/platinum/bts-hotcold.mdx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+## Explanation
+
+Let $a$, $b$, and $t$ be the nodes given in each query, and $p$ be
+the LCA of $a$ and $b$. Directly updating the path from $a$
+to $b$ is difficult to do, so breaking up the path into smaller segments
+is necessary. To do that, we use some casework.
+
+**Case 1: $p$ is equal to $t$, or $t$ is not inside $p$'s subtree.**
+
+In this case, we just update the path from $a$ to $p$ and
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:17]:
```suggestion
In this case, we update the path from $a$ to $p$ and
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/platinum/bts-hotcold.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/platinum/bts-hotcold.mdx]: done

File: solutions/platinum/bts-hotcold.mdx [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	      depth(n), tin(n), tout(n), diff(n) {
+		tin[0] = -1;
+		tout[0] = n + 1;
+		// ^ ensures that LCA works
+	}
+	void addEdge(int u, int v) {
+		adj[u].push_back(v);
+		adj[v].push_back(u);
+	}
+	void build(int u, int p) {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:77]: why not put this in the constructor

File: solutions/platinum/bts-hotcold.mdx [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+class Tree {
+  private:
+	const int n, LOG;
+	vector<vector<int>> adj, lift;
+	vector<int> depth, tin, tout;
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:61]: 1. declarations on separate lines
2. at this point just rename `LOG` to `log`

File: solutions/platinum/bts-hotcold.mdx [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	vector<array<ll, 2>> diff;
+	int timer = 0;
+
+  public:
+	Tree(int _n)
+	    : n(_n), LOG((int)log2(n) + 1), adj(n), lift(LOG, vector<int>(n)),
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:67]: putting `n` as the parameter is fine

File: solutions/platinum/bts-hotcold.mdx [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+		depth[u] = depth[p] + 1;
+		for (int i = 1; i < LOG; i++) {
+			lift[i][u] = lift[i - 1][lift[i - 1][u]];
+		}
+		for (int v : adj[u]) {
+			if (v != p) { build(v, u); }
+		}
+		tout[u] = timer - 1;
+	}
+	bool isAncestor(int u, int v) {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:89]: snake_case for names

File: solutions/platinum/bts-hotcold.mdx [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+		return tin[u] <= tin[v] && tout[v] <= tout[u];
+	}
+	int lca(int u, int v) {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:92]: newline between functions

File: solutions/platinum/bts-hotcold.mdx [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+	void dfs(int u, int p) {
+		for (int v : adj[u]) {
+			if (v == p) { continue; }
+			dfs(v, u);
+			diff[u][0] += diff[v][0] + diff[v][1];
+			diff[u][1] += diff[v][1];
+		}
+	}
+	vector<ll> result() {
+		dfs(1, 0);
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:162]: perhaps add a comment for what this does
or give a better name

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
YOU GET ME SO MUCH HOLY CRAP

File: solutions/platinum/bts-hotcold.mdx (hunk 8 above)
Comment (TheGamingMousse) [reply solutions/platinum/bts-hotcold.mdx:61]: std::log is a thing so ill prob name it smth diff

Comment (SansPapyrus683) [reply solutions/platinum/bts-hotcold.mdx]: this is why you don't using namespace std haha

File: solutions/platinum/bts-hotcold.mdx [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+class Tree {
+  private:
+	const int n, log2dist;
+	vector<vector<int>> &adj;
+	vector<vector<int>> lift;
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:62]:
```suggestion
	vector<vector<int>> lift;  // for binary lifting
```

File: solutions/platinum/bts-hotcold.mdx [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+	vector<array<ll, 2>> diff;
+	vector<int> depth;
+	vector<int> tin;
+	vector<int> tout;
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:66]: these are all for the euler tour right? put a comment on top "these 3 are for the euler tour"

File: solutions/platinum/bts-hotcold.mdx [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+class Tree {
+  private:
+	const int n, log2dist;
+	vector<vector<int>> &adj;
+	vector<vector<int>> lift;
+	vector<array<ll, 2>> diff;
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:63]: what this do

File: solutions/platinum/bts-hotcold.mdx [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+	vector<int> tout;
+	int timer = 0;
+
+  public:
+	Tree(int n, vector<vector<int>> &adj)
+	    : n(n), log2dist((int)log2(n) + 1), adj(adj),
+	      lift(log2dist, vector<int>(n)), depth(n), tin(n), tout(n), diff(n) {
+		tin[0] = -1;
+		tout[0] = n + 1;
+		// ^ ensures that LCA works
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:75]:
```suggestion
		tin[0] = -1;
		tout[0] = n + 1;  // ensures that LCA works
```

File: solutions/platinum/bts-hotcold.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/platinum/bts-hotcold.mdx:79]: bump?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
almost

File: solutions/platinum/bts-hotcold.mdx (hunk 7 above)
Comment (TheGamingMousse) [reply solutions/platinum/bts-hotcold.mdx:78]: the build() function? i can write that as a lambda if u want rq

Comment (SansPapyrus683) [reply solutions/platinum/bts-hotcold.mdx:80]: ok uh i took a closer look and found the name a bit misleading haha
hb a name like "tour" or "parse"?

Comment (TheGamingMousse) [reply solutions/platinum/bts-hotcold.mdx:80]: lol aight gimme a sec

Comment (SansPapyrus683) [reply solutions/platinum/bts-hotcold.mdx:80]: also make it private

File: solutions/platinum/bts-hotcold.mdx [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+	int lca(int u, int v) const {
+		if (is_ancestor(u, v)) { return u; }
+		if (is_ancestor(v, u)) { return v; }
+		for (int i = log2dist - 1; i >= 0; i--) {
+			if (!is_ancestor(lift[i][u], v)) { u = lift[i][u]; }
+		}
+		return lift[0][u];
+	}
+
+	int dist(int u, int v, int anc = -1) const {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx:115]: add a quick comment- "calculates the distance from u to anc and then to v (default is lca)"

**Review by SansPapyrus683 — APPROVED** [review]
alright basically done
just show ss of the new code ac-ing

**Comment by TheGamingMousse** [comment]
> alright basically done just show ss of the new code ac-ing

image

File: solutions/platinum/bts-hotcold.mdx [hunk 18]
```diff
@@ -99,11 +90,11 @@ class Tree {
 		tour(1, 0);
 	}
 
-	bool is_ancestor(int u, int v) const {
+	bool is_ancestor(int u, int v) {
```
Comment (SansPapyrus683) [inline solutions/platinum/bts-hotcold.mdx]: did you uh revert my changes :skull:

Comment (TheGamingMousse) [reply solutions/platinum/bts-hotcold.mdx]: wait what

uh that was an accident, lemme fix that

**Review by SansPapyrus683 — APPROVED** [review]
alright made some more changes
can i see ac again haha this'll be the last one i prommy

**Comment by TheGamingMousse** [comment]
> alright made some more changes can i see ac again haha this'll be the last one i prommy

image

**Review by ryanchou-dev — APPROVED** [review]
good!

**Pull request merged**